trio/tests/test_file_io.py

```diff
@@ -0,0 +1,169 @@
+import io
+import tempfile
+
+import pytest
+from unittest import mock
+from unittest.mock import patch, sentinel
+
+import trio
+from trio import _core
+from trio._file_io._file_io import _FILE_SYNC_ATTRS, _FILE_ASYNC_METHODS
+
+
+@pytest.fixture
+def path(tmpdir):
+    return tmpdir.join('test').__fspath__()
+
+
+@pytest.fixture
+def wrapped():
+    return mock.Mock(spec_set=io.StringIO)
+
+
+@pytest.fixture
+def async_file(wrapped):
+    return trio.wrap_file(wrapped)
+
+
+def test_wrap_invalid():
+    with pytest.raises(TypeError):
+        trio.wrap_file(str())
+
+
+def test_wrapped_property(async_file, wrapped):
+    assert async_file.wrapped == wrapped
+
+
+def test_dir_matches_wrapped(async_file, wrapped):
+    attrs = _FILE_SYNC_ATTRS + _FILE_ASYNC_METHODS
+
+    # all supported attrs in wrapped should be available in async_file
+    assert all(attr in dir(async_file) for attr in attrs if attr in dir(wrapped))
+    # all supported attrs not in wrapped should not be available in async_file
+    assert not any(attr in dir(async_file) for attr in attrs if attr not in dir(wrapped))
+
+
+def test_unsupported_not_forwarded(async_file):
+    async_file._wrapped = sentinel
+
+    assert hasattr(async_file.wrapped, 'unsupported_attr')
+
+    with pytest.raises(AttributeError):
+        getattr(async_file, 'unsupported_attr')
```

Comment:
I don't understand how this test works either. What's the advantage of using a 1-line fixture for `wrap_file(mock_object)` if you're then going to spend a line violating the abstraction boundary to convert that into `wrap_file(sentinel)`? And what magic properties are we relying on here -- `sentinel` objects have all attributes? I feel like this would be clearer if written like:

```python
async def test_unsupported_not_forwarded():
    class FakeFile:
        def write(self, data):
            pass
        def some_other_method(self):
           pass
    wrapper = wrap_file(FakeFile())
    assert hasattr(wrapper.wrapped, "write")
    assert hasattr(wrapper.wrapped, "some_other_method")
    assert hasattr(wrapper, "write")
    assert not hasattr(wrapper, "some_other_method")
    assert not hasattr(wrapper, "read")
```
...actually I guess I just rewrote part of the previous test in there too :-)